### ISSUE TITLE
Gracefully handle indentation whitespace for FTL format strings

### ISSUE THREAD

**Issue by Ardhimas** [projectfluent/fluent.js#127]
I spent a couple of hours today trying to figure out what was wrong with my code. I finally discovered that the issue is that the parser for FTL formatted strings does not gracefully handle indentation. For example:

```
const types = new MessageContext('en-US');
types.addMessages(`
  sets = {$count ->
    [one]   set
   *[other] sets
  }
  repeats = {$count ->
    [one]   repeats
   *[other] repeats
  }
`);
```

Throws the error `Expected an entry to start at the beginning of the file or on a new line`. It took me quite some time to figure out the error was in the whitespace, not in the newlines.

On the other hand, this works at the cost of poorer readability:

```
const types = new MessageContext('en-US');
types.addMessages(`
sets = {$count ->
[one]   set
*[other] sets
}
repeats = {$count ->
[one]   repeats
*[other] repeats
}
`);
```

I understand that this may or may not follow the FTL specs, but I think any developer will agree that improved readability through indentation is a good thing.

The [`react-native-l20n` package](https://github.com/jamesreggio/react-native-l20n) by @jamesreggio works perfectly in this regard if anyone would like to see an example of how it is implemented.

I'm curious to hear everyone's thoughts, thanks.

**Comment by stasm** [comment]
Thank you for taking the time to file this issue. I agree that there's room for improvement as far as the developer experience is involved in writing code like this.

There's a number of point that I'd like to make. I'll start with a minor one and move on to the more important ones.

1. `MessageContext.addMessages` shouldn't throw.

    > Throws the error Expected an entry to start at the beginning of the file or on a new line. It took me quite some time to figure out the error was in the whitespace, not in the newlines.

    Just to make it clear, do you mean that the error was _thrown_ or that it was present in the array of errors returned by `addMessages` ? It is important for our error recovery story that MessageContext not throw, so I'd like to make sure there isn't another bug here.

2. In current versions of Fluent, all of the message body must be indented, including the closing `}`.

    The snippet from you example shouldn't parse and indeed [it doesn't](http://projectfluent.org/play/?gist=7382904bd7cd3e0141664401ae61ef4f) in the Playground. 

    ```properties
    # Syntax error in current versions of Fluent.
    sets = {$count ->
      [one]   set
     *[other] sets
    }
    ```

    For some reason, the runtime parser used by `MessageContext` doesn't seem to care about the lack of the indentation. @zbraniecki would you mind taking a look?

    There are plans to relax the indentation requirement in the future. One change has already been accepted in the to the Syntax Spec (https://github.com/projectfluent/fluent/issues/42) but it hasn't been implemented in `fluent.js` yet. It will make it possible to put the closing `}` in the first column of the new line, without indentation.

3. The no-identation snippet shouldn't have worked.

    Your second snippet also shouldn't have worked, for the same reason as above: the entire body of the message must be indented. The parser which we use on Playground (the tooling parser) correctly [reports syntax errors](http://projectfluent.org/play/?gist=c7c13f5859bb5026f26d91d4d2c45973). @zbraniecki this again looks like a bug in the runtime parser.

4. How can we improve the error message?

    > It took me quite some time to figure out the error was in the whitespace, not in the newlines.

    Do you have a suggestion how we could improve the error message to make this easier for other people? Maybe simply `Expected an entry to start at the beginning of the line.`?

5. Fluent is white-space sensitive.

    Indentation matters in Fluent. This makes it a little bit harder to use FTL verbatim in JavaScript, Python or other programming languages where you'd like to indent FTL to match the code around it. It's a trade-off: the syntax doesn't need delimiters but it's optimized for being used in a separate file rathern than being embedded in a file written in another language. It's a common problem in many languages and the reason why utils like Python's [`textwrap.dedent`](https://docs.python.org/2/library/textwrap.html) exists.

    I wrote a similar [util function](https://github.com/projectfluent/fluent.js/blob/4db2bfbf602a9f112d099d76db1470502a66c413/fluent/test/util.js#L12-L22) for FTL. We use it for tests. Perhaps it would make sense to expose it as an export of the `fluent` package?

    I see that `react-native-l20n` also has [something similar](https://github.com/jamesreggio/react-native-l20n/blob/f4d79745bba12a1a40801ccc7006d43ec578c686/index.js#L77-L87) although it looks like it was written for the old syntax and it also does a bit more stripping.

    Sometimes in our own code we don't use any utils at all. The FTL then needs to [start in the first column of the line](https://github.com/projectfluent/fluent-rs/blob/85914d8f0af247405780dfb8ac49afd878eddaf0/tests/resolve_value.rs#L11). It looks weird at first but it makes sense and I personally got used to it.

**Comment by Ardhimas** [comment]
>Just to make it clear, do you mean that the error was thrown or that it was present in the array of errors returned by addMessages ? It is important for our error recovery story that MessageContext not throw, so I'd like to make sure there isn't another bug here.

It was not _thrown_, I was simply using that as a colloquial metaphor. I apologize for the confusion

>For some reason, the runtime parser used by MessageContext doesn't seem to care about the lack of the indentation. @zbraniecki would you mind taking a look?

>There are plans to relax the indentation requirement in the future. One change has already been accepted in the to the Syntax Spec (projectfluent/fluent#42) but it hasn't been implemented in fluent.js yet. It will make it possible to put the closing } in the first column of the new line, without indentation.

Thank you for clarifying. 

>Your second snippet also shouldn't have worked, for the same reason as above: the entire body of the message must be indented. The parser which we use on Playground (the tooling parser) correctly reports syntax errors. @zbraniecki this again looks like a bug in the runtime parser.

I actually accidentally removed the indentation when I posted the second example. It is actually equal to the first example, just without the initial block indentation:

```
sets = {$count ->
  [one]   set
 *[other] sets
}
```

Which is equal to the example you posted, so it's not 2 different bugs. Again, I apologize.

>Do you have a suggestion how we could improve the error message to make this easier for other people? Maybe simply Expected an entry to start at the beginning of the line.?

That sounds great to me. The confusion I ran into was in regards to the error `Expected an entry to start at the beginning of the file or on a new line` implying that the whitespace was not the issue, but the newline is.

>Indentation matters in Fluent. This makes it a little bit harder to use FTL verbatim in JavaScript, Python or other programming languages where you'd like to indent FTL to match the code around it. It's a trade-off: the syntax doesn't need delimiters but it's optimized for being used in a separate file rathern than being embedded in a file written in another language. It's a common problem in many languages and the reason why utils like Python's textwrap.dedent exists.

>I wrote a similar util function for FTL. We use it for tests. Perhaps it would make sense to expose it as an export of the fluent package?

>I see that react-native-l20n also has something similar although it looks like it was written for the old syntax and it also does a bit more stripping.

>Sometimes in our own code we don't use any utils at all. The FTL then needs to start in the first column of the line. It looks weird at first but it makes sense and I personally got used to it.

I can see 3 ways to do this:

- Exporting the util function
- Having an option such as `dedent: true` 
- Infer when there is a constant amount of initial whitespace and automatically sanitize the input

It looks to me like the util file does exactly what number 3 does, so if it doesn't break anything to run everything through the function, I don't see a reason not to do so.

Thank you for the in-depth response. I hope my suggestions are useful to you.

**Comment by stasm** [comment]
Thanks so much, it's great to have this discussion!

The parsing done in `addMessages` is one of the performance bottlenecks of `MessageContext`. For this reason I'd like to avoid automatically dedenting the input.

As far as the proposed `dedent: true` option goes, I feel like the primary use-case of `addMessages` is to pass FTL code which was fetched from a file (without any indentation) rather than hard-coded in the source. In fact, hard-coding FTL in the source is most likely only useful for very small apps, prototypes and for testing. I don't think extending the API of `addMessages` is justified by these use-cases.

The above use-cases can be very well served by the `ftl` dedent helper, however. I decided to go ahead and expose it from the `fluent` package: https://github.com/projectfluent/fluent.js/commit/24585e681edb28c04f0d43e8ceb6621205803def. It will be available in the next version of `fluent`, 0.6, which I plan to release this week.

**Comment by Ardhimas** [comment]
@stasm Thank you for everything. I'll close this issue now as it's resolved pending the next PR.
